clients/githubrepo/branches.go

```diff
@@ -143,12 +253,51 @@ func (handler *branchesHandler) setup() error {
 			"owner": githubv4.String(handler.repourl.owner),
 			"name":  githubv4.String(handler.repourl.repo),
 		}
+
+		rulesData := new(ruleSetData)
+		if err := handler.graphClient.Query(handler.ctx, rulesData, vars); err != nil {
+			handler.errSetup = sce.WithMessage(sce.ErrScorecardInternal, fmt.Sprintf("githubv4.Query: %v", err))
+			return
+		}
+		handler.ruleSets = make([]*repoRuleSet, 0)
+		for _, rule := range rulesData.Repository.Rulesets.Nodes {
+			if rule.Enforcement == nil || *rule.Enforcement != "ACTIVE" {
+				continue
+			}
+			handler.ruleSets = append(handler.ruleSets, rule)
+		}
+
 		handler.data = new(defaultBranchData)
 		if err := handler.graphClient.Query(handler.ctx, handler.data, vars); err != nil {
-			handler.errSetup = sce.WithMessage(sce.ErrScorecardInternal, fmt.Sprintf("githubv4.Query: %v", err))
+			// If the repository is using RuleSets, ignore permissions errors on branch protection as that requires admin.
+			if len(handler.ruleSets) == 0 || !isPermissionsError(err) {
+				handler.errSetup = sce.WithMessage(sce.ErrScorecardInternal, fmt.Sprintf("githubv4.Query: %v", err))
+				return
+			}
+
+			// To recover, we still need to know the default branch name:
+			defaultBranchNameData := new(defaultBranchNameData)
+			if err := handler.graphClient.Query(handler.ctx, defaultBranchNameData, vars); err != nil {
```

Comment:
I really like the results of this change.

I wondered about renaming the query types to reflect their expected access level instead of their data:
Currently I have a comment that `ruleSetData` is fetched before `defaultBranchData` because it will be accessible to tokens with lesser scope. I thought renaming those types  to something like `publicBranchData` and `adminBranchData` might further encode that difference.

Reply:
I took a look at our [documentation around Branch-Protection](https://github.com/ossf/scorecard/blob/accaffdfe2aacc8bc48a15e375f1dc9c338b3655/docs/checks.md#branch-protection) and every setting marked as `For administrators`, is accessible to repo rules (with the distinction of the exact bypass actors (see our other discussion)).

Given that the data is more accessible, and GitHub refers to repo rules as "GitHub’s next evolution of branch protections", I think your current approach of focusing on repo rules first makes sense. We'll likely refine our documentation in a follow-up PR to start advocating for repo rules instead, and avoiding the issue of PAT in the action.

> I thought renaming those types to something like publicBranchData and adminBranchData might further encode that difference.

What do you think about framing it instead as data source difference, e.g. `ruleSetData` vs `branchProtectionData`?